WalletWasabi/Services/SingleInstanceChecker.cs

```diff
@@ -1,95 +1,156 @@
+using Microsoft.Extensions.Hosting;
 using NBitcoin;
-using Nito.AsyncEx;
 using System;
-using System.IO;
+using System.Net;
+using System.Net.Sockets;
 using System.Threading;
 using System.Threading.Tasks;
+using WalletWasabi.Logging;
 
 namespace WalletWasabi.Services
 {
-	public class SingleInstanceChecker : IDisposable
+	public class SingleInstanceChecker : BackgroundService, IAsyncDisposable
 	{
-		/// <summary>Unique prefix for global mutex name.</summary>
-		private const string MutexString = "WalletWasabiSingleInstance";
-
-		/// <summary>Name of system-wide mutex.</summary>
-		private readonly string _lockName;
-
-		private bool _disposedValue;
-
 		/// <summary>
-		/// Creates a new instance of the object where lock name is based on <paramref name="network"/> name.
+		/// Creates an object to ensure mutual exclusion of Wasabi instances per Network <paramref name="network"/>.
+		/// The solution based on TCP socket.
 		/// </summary>
-		/// <param name="network">Bitcoin network selected when Wasabi Wallet was started.</param>
-		public SingleInstanceChecker(Network network) : this(network, network.ToString())
+		/// <param name="network">Bitcoin network selected when Wasabi Wallet was started. It will use the port 37129,37130,37131 according to network main,test,reg.</param>
+		public SingleInstanceChecker(Network network) : this(NetworkToPort(network))
 		{
 		}
 
 		/// <summary>
 		/// Use this constructor only for testing.
 		/// </summary>
-		/// <param name="network">Bitcoin network selected when Wasabi Wallet was started.</param>
-		/// <param name="lockName">Name of system-wide mutex.</param>
-		public SingleInstanceChecker(Network network, string lockName)
+		public SingleInstanceChecker(int port)
 		{
-			Network = network;
-			_lockName = $"{MutexString}-{lockName}";
+			Port = port;
 		}
 
-		private IDisposable? SingleApplicationLockHolder { get; set; }
-		private Network Network { get; }
+		private int Port { get; }
 
-		public async Task CheckAsync()
+		private CancellationTokenSource DisposeCts { get; } = new();
+		private TaskCompletionSource? TaskStartTcpListener { get; set; }
+
+		public event EventHandler? OtherInstanceStarted;
+
+		/// <summary>
+		/// This function ensures this is the first instance running on this machine or throws an exception if we are not. In case of secondary start
+		/// we are trying signal the first instance before throwing the exception.
+		/// On macOS this funtion will never throw if you run Wasabi as a macApp, because mac prevents running the same app multiple time on OS level.
```

Comment:
```suggestion
		/// This function ensures that this is the first instance running on this machine or throws an exception if it is not. In case of secondary start
		/// we try to signal the first instance before throwing the exception.
		/// On macOS this function will never throw if you run Wasabi as a macApp, because mac prevents running the same app multiple times on OS level.
```